crates/primitives/src/checkpoints.rs

```diff
@@ -121,6 +123,88 @@ pub struct StorageHashingCheckpoint {
     pub to: u64,
 }
 
+/// Saves the progress of a stage.
+#[main_codec]
+#[derive(Debug, Default, PartialEq, Eq, Clone, Copy)]
+pub struct StageCheckpoint {
+    /// The maximum block processed by the stage.
+    pub block_number: BlockNumber,
+    /// Stage-specific checkpoint. None if stage uses only block-based checkpoints.
+    pub stage_checkpoint: Option<StageUnitCheckpoint>,
+}
+
+impl StageCheckpoint {
+    /// Creates a new [`StageCheckpoint`] with only `block_number` set.
+    pub fn new_with_block_number(block_number: BlockNumber) -> Self {
+        Self { block_number, ..Default::default() }
+    }
+}
```

Comment:
having constructor functions for this type is useful, because this is called a lot.

I'd suggest renaming this to just `new` and then `new_with_checkpoint` or something

Reply:
agree on `new`, for checkpoint units plan to do something like this:
```rust
    /// Returns the account hashing stage checkpoint, if any.
    pub fn account_hashing_stage_checkpoint(&self) -> Option<AccountHashingCheckpoint> {
        match self.stage_checkpoint {
            Some(StageUnitCheckpoint::Account(checkpoint)) => Some(checkpoint),
            _ => None,
        }
    }

    /// Sets the stage checkpoint to account hashing.
    pub fn with_account_hashing_stage_checkpoint(
        mut self,
        checkpoint: AccountHashingCheckpoint,
    ) -> Self {
        self.stage_checkpoint = Some(StageUnitCheckpoint::Account(checkpoint));
        self
    }
```

Reply:
I likes this!